addons/controllers/tanzuclusterbootstrap_controller.go

```diff
@@ -0,0 +1,586 @@
+					r.providerGVR[gk] = &schema.GroupVersionResource{Group: gv.Group, Resource: apiResource.APIResources[i].Name, Version: gv.Version}
+					return r.providerGVR[gk], nil
+				}
+			}
+		}
+	}
+
+	return nil, fmt.Errorf("unable to find server preferred resource %s/%s", gk.Group, gk.Kind)
+}
+
+// periodicGVRCachesClean invalidates caches used for GVR lookup
+func (r *TanzuClusterBootstrapReconciler) periodicGVRCachesClean() {
+	ticker := time.NewTicker(constants.DiscoveryCacheInvalidateInterval)
+	for {
+		select {
+		case <-r.context.Done():
+			ticker.Stop()
+			return
+		case <-ticker.C:
+			r.cachedDiscoveryClient.Invalidate()
+			r.providerGVR = make(map[schema.GroupKind]*schema.GroupVersionResource)
+		}
+	}
+}
+
+// updateValuesFromSecret updates secretRef in valuesFrom
+func (r *TanzuClusterBootstrapReconciler) updateValuesFromSecret(cluster *clusterapiv1beta1.Cluster, bootstrap *runtanzuv1alpha3.TanzuClusterBootstrap,
+	valuesFrom *runtanzuv1alpha3.ValuesFrom, templateNS string, packageShortName string, log logr.Logger) (*corev1.Secret, error) {
+
+	var newSecret *corev1.Secret
+	if valuesFrom.SecretRef != "" {
+		secret := &corev1.Secret{}
+		key := client.ObjectKey{Namespace: templateNS, Name: valuesFrom.SecretRef}
+		if err := r.Get(r.context, key, secret); err != nil {
+			log.Error(err, "unable to fetch secret", "objectkey", key)
+			return nil, err
+		}
+		newSecret = secret.DeepCopy()
+		newSecret.OwnerReferences = []metav1.OwnerReference{
+			{
+				APIVersion: clusterapiv1beta1.GroupVersion.String(),
+				Kind:       cluster.Kind,
+				Name:       cluster.Name,
+				UID:        cluster.UID,
+			},
+		}
+		newSecret.Name = fmt.Sprintf("%s-%s-package", cluster.Name, packageShortName)
+		newSecret.Namespace = bootstrap.Namespace
+		if err := r.Create(r.context, newSecret); err != nil {
+			return nil, err
+		}
+		valuesFrom.SecretRef = newSecret.Name
+	}
+
+	return newSecret, nil
+}
+
+// updateValuesFromProvider updates providerRef in valuesFrom
+func (r *TanzuClusterBootstrapReconciler) updateValuesFromProvider(cluster *clusterapiv1beta1.Cluster, bootstrap *runtanzuv1alpha3.TanzuClusterBootstrap,
+	valuesFrom *runtanzuv1alpha3.ValuesFrom, templateNS string, packageShortName string, log logr.Logger) (*unstructured.Unstructured, error) {
+
+	var newProvider *unstructured.Unstructured
+	if valuesFrom.ProviderRef != nil {
+		gvr, err := r.getGVR(schema.GroupKind{Group: *valuesFrom.ProviderRef.APIGroup, Kind: valuesFrom.ProviderRef.Kind})
+		if err != nil {
+			log.Error(err, "failed to getGVR")
+			return nil, err
+		}
+		provider, err := r.dynamicClient.Resource(*gvr).Namespace(templateNS).Get(r.context, valuesFrom.ProviderRef.Name, metav1.GetOptions{})
+		if err != nil {
+			log.Error(err, "unable to fetch provider", "provider", valuesFrom.ProviderRef, "gvr", gvr)
+			return nil, err
+		}
+		newProvider = provider.DeepCopy()
+		unstructured.RemoveNestedField(newProvider.Object, "metadata")
```

Comment:
question: This line makes me think if we want to completely remove `metadata` from `newProvider` OR we could skip this line to preserve something like `annotations` under metadata ? I believe the code below still overwrites `OwnerRef`, `Name` and `Namespace`.

Reply:
Not just annotations, there could be labels as well and they probably don't mean anything for cloned objects. By clearing out we aren't making any implicit guarantees for consumers relying on annotations if any. I am thinking this could be improved by doing a deepCopy of just the spec field instead since that's what I want.